File: tobas_property/tobas_property_tools_py/tobas_property_tools_py/property_client.py
import os.path as osp
from rclpy.node import Node
from typing import Tuple, Type, Any
from std_srvs.srv import Trigger

from tobas_property_msgs.srv import GetBool, GetInt, GetDouble, GetString, SetBool, SetInt, SetDouble, SetString


class PropertyClient:
    # Error Code
    E_NO_ERROR = 0
    E_SERVICE_NOT_READY = -1
    E_FAILED_TO_CALL = -2
    E_SERVER_ERROR = -3

    GET_BOOL_SRV = "get_bool"
    GET_INT_SRV = "get_int"
    GET_DOUBLE_SRV = "get_double"
    GET_STRING_SRV = "get_string"
    SET_BOOL_SRV = "set_bool"
    SET_INT_SRV = "set_int"
    SET_DOUBLE_SRV = "set_double"
    SET_STRING_SRV = "set_string"
    SAVE_FILE_SRV = "save_file"

    def __init__(self, node: Node, ns: str, section: str) -> None:
        self._node = node
        self._ns = ns
        self._section = section

        self._error_code = self.E_NO_ERROR
        self._server_error_msg = ""

    def get_bool(self, key: str) -> Tuple[int, bool | None]:
        return self._get_property(key, self.GET_BOOL_SRV, GetBool)

    def get_int(self, key: str) -> Tuple[int, int | None]:
        return self._get_property(key, self.GET_INT_SRV, GetInt)

    def get_float(self, key: str) -> Tuple[int, float | None]:
        return self._get_property(key, self.GET_DOUBLE_SRV, GetDouble)

    def get_string(self, key: str) -> Tuple[int, str | None]:
        return self._get_property(key, self.GET_STRING_SRV, GetString)

    def set_bool(self, key: str, value: bool) -> int:
        return self._set_property(key, value, self.SET_BOOL_SRV, SetBool)

    def set_int(self, key: str, value: int) -> int:
        return self._set_property(key, value, self.SET_INT_SRV, SetInt)

    def set_float(self, key: str, value: float) -> int:
        return self._set_property(key, value, self.SET_DOUBLE_SRV, SetDouble)

    def set_string(self, key: str, value: str) -> int:
        return self._set_property(key, value, self.SET_STRING_SRV, SetString)

    def save(self) -> int:
        client = self._node.create_client(Trigger, osp.join(self._ns, self.SAVE_FILE_SRV))

        self._node.get_logger().debug(f'Waiting for "{self.SAVE_FILE_SRV}" service server.')
        if not client.service_is_ready():
            self._error_code = self.E_SERVICE_NOT_READY
            return self._error_code

        req = Trigger.Request()

        self._node.get_logger().debug(f'Calling "{self.SAVE_FILE_SRV}" service.')
        res: Trigger.Response = client.call(req)
        if not res.success:
            self._error_code = self.E_SERVER_ERROR
            self._server_error_msg = res.message
            return self._error_code

        self._error_code = self.E_NO_ERROR
        return self._error_code

    def error_code(self) -> int:
        return self._error_code

    def error_message(self) -> str:
        if self._error_code == self.E_NO_ERROR:
            return ""
        elif self._error_code == self.E_SERVICE_NOT_READY:
            return "Property server is not ready."
        elif self._error_code == self.E_FAILED_TO_CALL:
            return "Failed to call property service."
        elif self._error_code == self.E_SERVER_ERROR:
            return self._server_error_msg
        else:
            raise

    def _get_property(self, key: str, srv_name: str, SrvType: Type) -> Tuple[int, Any]:
        client = self._node.create_client(SrvType, osp.join(self._ns, srv_name))

        self._node.get_logger().debug(f'Waiting for "{srv_name}" service server.')
        if not client.service_is_ready():
            self._error_code = self.E_SERVICE_NOT_READY
            return self._error_code

        req = SrvType.Request()
        req.section = self._section
        req.key = key

        self._node.get_logger().debug(f'Calling "{srv_name}" service.')
        res: SrvType.Response = client.call(req)
        if not res.success:
            self._error_code = self.E_SERVER_ERROR
            self._server_error_msg = res.message
            return self._error_code, None

        self._error_code = self.E_NO_ERROR
        return self._error_code, res.value

    def _set_property(self, key: str, value: Any, srv_name: str, SrvType: Type) -> int:
        client = self._node.create_client(SrvType, osp.join(self._ns, srv_name))

        self._node.get_logger().debug(f'Waiting for "{srv_name}" service server.')
        if not client.service_is_ready():
            self._error_code = self.E_SERVICE_NOT_READY
            return self._error_code

        req = SrvType.Request()
        req.section = self._section
        req.key = key
        req.value = value

        self._node.get_logger().debug(f'Calling "{srv_name}" service.')
        res: SrvType.Response = client.call(req)
        if not res.success:
            self._error_code = self.E_SERVER_ERROR
            self._server_error_msg = res.message
            return self._error_code

        self._error_code = self.E_NO_ERROR
        return self._error_code

File: tobas_property/tobas_property_client/src/property_client.cpp
#include <std_srvs/srv/trigger.hpp>

#include <tobas_property_msgs/srv/get_bool.hpp>
#include <tobas_property_msgs/srv/get_int.hpp>
#include <tobas_property_msgs/srv/get_double.hpp>
#include <tobas_property_msgs/srv/get_string.hpp>
#include <tobas_property_msgs/srv/set_bool.hpp>
#include <tobas_property_msgs/srv/set_int.hpp>
#include <tobas_property_msgs/srv/set_double.hpp>
#include <tobas_property_msgs/srv/set_string.hpp>

#include "../include/tobas_property_client/property_client.hpp"

using namespace std;
using namespace std_srvs::srv;
using namespace tobas_property_msgs::srv;

namespace ptree
{
PropertyClient::PropertyClient(rclcpp::Node::SharedPtr node, const string& ns, const string& section)
  : node_(node), ns_(ns), section_(section)
{
}

PropertyClient::error_t PropertyClient::get(const string& key, bool& value)
{
  return getProperty<GetBool, kGetBoolSrv>(key, value);
}

PropertyClient::error_t PropertyClient::get(const string& key, int& value)
{
  return getProperty<GetInt, kGetIntSrv>(key, value);
}

PropertyClient::error_t PropertyClient::get(const string& key, double& value)
{
  return getProperty<GetDouble, kGetDoubleSrv>(key, value);
}

PropertyClient::error_t PropertyClient::get(const string& key, string& value)
{
  return getProperty<GetString, kGetStringSrv>(key, value);
}

PropertyClient::error_t PropertyClient::get(const string& key, uint8_t& value)
{
  int tmp;

  if (getProperty<GetInt, kGetIntSrv>(key, tmp) < 0)
    return error_code_;

  if (tmp < 0 || UINT8_MAX < tmp)
    return error_code_ = E_OUT_OF_RANGE;

  value = static_cast<uint8_t>(tmp);
  return error_code_ = E_NO_ERROR;
}

PropertyClient::error_t PropertyClient::get(const string& key, uint16_t& value)
{
  int tmp;

  if (getProperty<GetInt, kGetIntSrv>(key, tmp) < 0)
    return error_code_;

  if (tmp < 0 || UINT16_MAX < tmp)
    return error_code_ = E_OUT_OF_RANGE;

  value = static_cast<uint16_t>(tmp);
  return error_code_ = E_NO_ERROR;
}

PropertyClient::error_t PropertyClient::get(const string& key, float& value)
{
  return getProperty<GetDouble, kGetDoubleSrv>(key, value);
}

PropertyClient::error_t PropertyClient::set(const string& key, const bool& value)
{
  return setProperty<SetBool, kSetBoolSrv>(key, value);
}

PropertyClient::error_t PropertyClient::set(const string& key, const int& value)
{
  return setProperty<SetInt, kSetIntSrv>(key, value);
}

PropertyClient::error_t PropertyClient::set(const string& key, const double& value)
{
  return setProperty<SetDouble, kSetDoubleSrv>(key, value);
}

PropertyClient::error_t PropertyClient::set(const string& key, const string& value)
{
  return setProperty<SetString, kSetStringSrv>(key, value);
}

PropertyClient::error_t PropertyClient::set(const string& key, const uint8_t& value)
{
  return setProperty<SetInt, kSetIntSrv>(key, value);
}

PropertyClient::error_t PropertyClient::set(const string& key, const uint16_t& value)
{
  return setProperty<SetInt, kSetIntSrv>(key, value);
}

PropertyClient::error_t PropertyClient::set(const string& key, const float& value)
{
  return setProperty<SetDouble, kSetDoubleSrv>(key, value);
}

PropertyClient::error_t PropertyClient::save()
{
  const auto client = node_->create_client<Trigger>(path::join(ns_, kSaveFileSrv));
  if (!client->service_is_ready())
    return error_code_ = E_SERVICE_NOT_READY;

  const auto req = make_shared<Trigger::Request>();

  auto future = client->async_send_request(req);
  future.wait();

  const auto res = future.get();
  if (!res->success)
  {
    server_error_msg_ = res->message;
    return error_code_ = E_SERVER_ERROR;
  }

  return error_code_ = E_NO_ERROR;
}

PropertyClient::error_t PropertyClient::errorCode() const
{
  return error_code_;
}

const char* PropertyClient::errorMessage() const
{
  switch (error_code_)
  {
    case E_NO_ERROR:
      return "";
    case E_SERVICE_NOT_READY:
      return "Property server is not ready.";
    case E_OUT_OF_RANGE:
      return "The property value is out of numerical range.";
    case E_SERVER_ERROR:
      return server_error_msg_.c_str();
    default:
      throw;
  }
}
}  // namespace ptree
